### PR TITLE
optim: vector table moved to RAM on STM32 MCUs

### PR DESCRIPTION
**Pull request by cruwaller**
1K of RAM reserved for code now. And ISR vector table linked into RAM to reduce FLASH accesses.

### CODE REVIEW

**Review by DoghouseFPV — COMMENTED** [review]
This appears to be moving some critical ISR code into RAM.   Does it also move the vector table itself into RAM?   If not, that addition could also be of benefit to reduce flash accesses.

Also, the VTOR register is, I thought, supposed to be an offset....not an absolute address.  This line:
SCB->VTOR = (__IO uint32_t) &g_pfnVectors;
...seems to be treating it as an absolute address.  I think in some cases, it ends up not mattering because of the memory mapped modes on STM32.  So it may appear to work even if the code is mis-leading by not passing in an offset.
"VTOR" = Vector Table Offset Register

**Comment by cruwaller** [comment]
This actually just moves the ISR table into RAM.
No other code yet...

VTOR name is a bit misleading but it has to be the base address of vector table :)

**Comment by cruwaller** [comment]
From ARM developer site:

> The VTOR indicates the offset of the vector table base address from memory address 0x00000000

**Comment by cruwaller** [comment]
@pkendall64 I rebased this PR to latest master. I haven't tested yet but just FYI if you are interested of this...

**Comment by pkendall64** [comment]
> @pkendall64 I rebased this PR to latest master. I haven't tested yet but just FYI if you are interested of this...

I'll check it out and see if works. Cheers.

File: src/targets/common.ini [hunk 1]
```diff
@@ -40,6 +40,7 @@ build_flags =
 	-D PLATFORM_ESP32=1
 	-D CONFIG_TCPIP_LWIP=1
 	-D BEARSSL_SSL_BASIC
+	-I ${PROJECTSRC_DIR}/hal
```
Comment (pkendall64) [inline src/targets/common.ini:43]: Not needed

File: src/targets/common.ini [hunk 2]
```diff
@@ -80,7 +82,10 @@ board = bluepill_f103c8
 build_unflags = -Os
 build_flags =
 	-D PLATFORM_STM32=1
+	-Wl,-Map,firmware.map
 	-O2
+	-I ${PROJECTSRC_DIR}/hal
```
Comment (pkendall64) [inline src/targets/common.ini:87]: Not needed

File: src/targets/common.ini [hunk 3]
```diff
@@ -57,6 +58,7 @@ build_flags =
 	-D CONFIG_TCPIP_LWIP=1
 	-D BEARSSL_SSL_BASIC
 	-O2
+	-I ${PROJECTSRC_DIR}/hal
```
Comment (pkendall64) [inline src/targets/common.ini:61]: Not needed

File: src/targets/common.ini [hunk 4]
```diff
@@ -80,7 +82,10 @@ board = bluepill_f103c8
 build_unflags = -Os
 build_flags =
 	-D PLATFORM_STM32=1
+	-Wl,-Map,firmware.map
 	-O2
+	-I ${PROJECTSRC_DIR}/hal
 src_filter = ${common_env_data.src_filter} -<ESP32*.*> -<ESP8266*.*> -<WS281B*.*>
 lib_deps =
 	paolop74/extEEPROM @ ^3.4.1
+	HAL
```
Comment (pkendall64) [inline src/targets/common.ini:91]: Not needed

**Comment by pkendall64** [comment]
I've rebuild all the firmware and verified the address of the vectors and the ones that are checked by the bootloader all fall within the range so it will pass and execute the firmware. I also flash the R9MM with the firmware and it works perfectly.
I think if those minor things I mentioned were fixed/removed we should get this approved and merged ASAP.

File: src/targets/common.ini (hunk 1 above)
Comment (cruwaller) [reply src/targets/common.ini:43]: Do we wanna remove these? I added these for future usage...

File: src/targets/common.ini (hunk 4 above)
Comment (cruwaller) [reply src/targets/common.ini:91]: oops, typo. Will be removed. thanks!

**Comment by cruwaller** [comment]
> I've rebuilt all the firmware and verified the address of the vectors and the ones that are checked by the bootloader all fall within the range so it will pass and execute the firmware.

Thanks!
I run also some tests with the receivers I have and all seem to work as expected. All STM based receivers need to be tested well with lto is enabled (PR #1169).

File: src/targets/common.ini (hunk 1 above)
Comment (pkendall64) [reply src/targets/common.ini:43]: Any sub-folders you add to the lib folder is automatically added to the include path.

**Comment by pkendall64** [comment]
> > I've rebuilt all the firmware and verified the address of the vectors and the ones that are checked by the bootloader all fall within the range so it will pass and execute the firmware.
> 
> Thanks! I run also some tests with the receivers I have and all seem to work as expected. All STM based receivers need to be tested well with lto is enabled (PR #1169).

My R9MM RX just will not work with lto enabled. I moved so serial debugging further up and nothing is printed after the bootloader, but the LEDs go out so I assume it's trying to run the code but failing.

File: src/targets/common.ini (hunk 1 above)
Comment (cruwaller) [reply src/targets/common.ini:43]: I tried to add that hal file into libs but it wont compile it without any header... I can remove these include paths, no problem :)

Comment (pkendall64) [reply src/targets/common.ini:43]: I just tested moving src/hal to lib/HAL and removing the `-I` but leaving the `HAL` library include and it compiles and links it in.

**Comment by cruwaller** [comment]
> > > I've rebuilt all the firmware and verified the address of the vectors and the ones that are checked by the bootloader all fall within the range so it will pass and execute the firmware.
> > 
> > 
> > Thanks! I run also some tests with the receivers I have and all seem to work as expected. All STM based receivers need to be tested well with lto is enabled (PR #1169).
> 
> My R9MM RX just will not work with lto enabled. I moved so serial debugging further up and nothing is printed after the bootloader, but the LEDs go out so I assume it's trying to run the code but failing.

Hmm, strange.. I didn't try lto yet. Need to attach the debugger and see what happens 🤔

File: src/targets/common.ini (hunk 4 above)
Comment (pkendall64) [reply src/targets/common.ini:91]: Yeah, I like the way you've done it actually, so just delete this one. Then I think it's good to go unless there are changes required for lto.

File: src/targets/common.ini (hunk 1 above)
Comment (cruwaller) [reply src/targets/common.ini:43]: Yeah, later I realize that stm common `lib_deps` is not include into all envs...

File: src/targets/common.ini (hunk 4 above)
Comment (cruwaller) [reply src/targets/common.ini]: fixed

**Review by pkendall64 — APPROVED** [review]

**Comment by StonedDawg** [comment]
I have tested this with R9M and R9 SLIMPLUS OTA, switching my config around and all is working fine

**Review by JyeSmith — APPROVED** [review]
Tested with Nam Flash to HM PP.  No obvious issues.

**Pull request merged**